### ISSUE TITLE
how do I turn a docker image into a sandstorm package?

### ISSUE THREAD

**Issue by dckc** [sandstorm-io/sandstorm#911]
I suspect the general case answer is: you don't. The architecture is different.

But perhaps it could work in some simple cases and we could demonstrate one or two under [Raw Packaging](https://docs.sandstorm.io/en/latest/developing/raw-packaging-guide/)?

**Comment by zarvox** [comment]
Yeah, in the general case, you don't.  There's a lengthier discussion on why not here: https://blog.sandstorm.io/news/2014-08-19-why-not-run-docker-apps.html

On the bright side, I'll note that when packaging existing apps, docker images (and in particular, the Dockerfile) can often be a helpful reference for how to build, package, and run the app in question.  The presence of a Dockerfile usually indicates that there's an automated build flow for producing a chroot that will run the app, which port it will listen on, what commands to run inside the container and so on.

These things tend to map pretty well to `vagrant-spk`'s concept of "setup.sh", "build.sh", and "launcher.sh" scripts:

* `setup.sh`: installing prerequisites.  This is where you'd set up your base image (`FROM <imagename>`) and install dependency packages (build toolchain, language tools, etc.)
* `build.sh`: building the app.  This is where you'd put the equivalent of `RUN ...` lines that build your application from source, or assemble an app bundle, or things like that.
* `launcher.sh`: running the app.  This is where you'd put any per-instance data setup, migrations, and starting the actual service.  This is akin to the `CMD` line in a Dockerfile.

There's a little more to it than what I wrote above - some apps will expect to connect to another container providing e.g. a database server, and on Sandstorm you'd be expected to put that in your own package, so your Sandstorm package provides a unit of functionality for some user, rather than some implementation detail of a service.  You can also possibly infer which port to tell `sandstorm-http-bridge` to listen on (rather than port 8000) from the `EXPOSE` directive.

But you can generally use a Dockerfile as a guide to get an app up and running inside Sandstorm pretty quickly. :)

**Comment by dckc** [comment]
The FROM->setup, RUN->build, CMD->launcher connection is just the sort of thing I had in mind. I'll keep an eye out for something that would make a good worked example.

And thanks for reminding me about the blog item.

**Comment by ocdtrekkie** [comment]
I think we should treat this in 2020 as a request for documentation on https://github.com/zenhack/docker-spk
